File: api/waitlist/skills.py
from flask import Blueprint, g
from . import auth
from .data import skills
from .data.database import Character, SkillHistory
from .webutil import ViewReturn
from .tdf import skills as tdf_skills

bp = Blueprint("skills", __name__)


@bp.route("/api/skills")
@auth.login_required
@auth.select_character(override_permission="skill-view")
def get_skills() -> ViewReturn:
    overview = {}

    character = g.db.query(Character).filter(Character.id == g.character_id).one()
    skill_lookup = skills.load_character_skills(g.character_id)

    for skill_id in tdf_skills.RELEVANT_SKILLS:
        overview[skill_id] = skill_lookup[skill_id] if skill_id in skill_lookup else 0

    return {
        "current": overview,
        "categories": tdf_skills.CATEGORIES,
        "requirements": tdf_skills.REQUIREMENTS,
        "ids": tdf_skills.SKILL_IDS,
        "character_name": character.name,
    }


@bp.route("/api/history/skills")
@auth.login_required
@auth.select_character(override_permission="skill-history-view")
def get_skills_history() -> ViewReturn:
    relevant = set(tdf_skills.RELEVANT_SKILLS)

    history = []

    for history_entry in (
        g.db.query(SkillHistory)
        .filter(SkillHistory.character_id == g.character_id)
        .order_by(SkillHistory.id.desc())
    ):
        if not history_entry.skill_id in relevant:
            continue

        history.append(
            {
                "skill_id": history_entry.skill_id,
                "old_level": history_entry.old_level,
                "new_level": history_entry.new_level,
                "logged_at": history_entry.logged_at,
            }
        )

    return {
        "history": history,
        "ids": tdf_skills.SKILL_IDS,
    }

File: backend/src/routes/history/skills.rs
use eve_data_core::{SkillLevel, TypeID};
use rocket::serde::json::Json;
use serde::Serialize;
use std::collections::HashMap;

use crate::tdf::skills as tdf_skills;
use crate::{
    core::auth::{authorize_character, AuthenticatedAccount},
    util::madness::Madness,
};

#[derive(Serialize)]
struct SkillHistoryResponseLine {
    skill_id: TypeID,
    old_level: SkillLevel,
    new_level: SkillLevel,
    logged_at: i64,
}

#[derive(Serialize)]
struct SkillHistoryResponse {
    history: Vec<SkillHistoryResponseLine>,
    ids: &'static HashMap<String, TypeID>,
}

#[get("/api/history/skills?<character_id>")]
async fn skill_history(
    app: &rocket::State<crate::app::Application>,
    account: AuthenticatedAccount,
    character_id: i64,
) -> Result<Json<SkillHistoryResponse>, Madness> {
    authorize_character(
        app.get_db(),
        &account,
        character_id,
        Some("skill-history-view"),
    )
    .await?;
    let relevance = &tdf_skills::skill_data().relevant_skills;

    let history = sqlx::query!(
        "SELECT * FROM skill_history WHERE character_id = ? ORDER BY id DESC",
        character_id
    )
    .fetch_all(app.get_db())
    .await?
    .into_iter()
    .map(|row| SkillHistoryResponseLine {
        skill_id: row.skill_id as TypeID,
        old_level: row.old_level as SkillLevel,
        new_level: row.new_level as SkillLevel,
        logged_at: row.logged_at,
    })
    .filter(|row| relevance.contains(&row.skill_id))
    .collect();

    Ok(Json(SkillHistoryResponse {
        history,
        ids: &tdf_skills::skill_data().name_lookup,
    }))
}

pub fn routes() -> Vec<rocket::Route> {
    routes![skill_history]
}
